src/tests/system/docs/concepts.rst

```diff
@@ -0,0 +1,113 @@
+    new, clear, documented and reusable API
+
+  * Avoid calling commands on remote hosts directly from the tests, this belongs
+    to :mod:`lib.sssd.roles`
+
+  * If you need to call a command, it most likely means that you want to extend
+    :mod:`lib.sssd.roles` module
+
+* Use topology parametrization whenever possible
+
+  * This rapidly increases the code coverage
+
+Naming tests
+************
+
+Name your tests as ``test_feature__case``. For example:
+
+.. code-block:: python
+
+    @pytest.mark.topology(KnownTopologyGroup.AnyProvider)
+    def test_id__shortname():
+        pass
+
+    @pytest.mark.topology(KnownTopologyGroup.AnyProvider)
+    def test_id__fqn():
+        pass
+
+    @pytest.mark.topology(KnownTopologyGroup.AnyProvider)
+    def test_id__name_with_space():
+        pass
+
+About using fixtures
+********************
+
+`Fixtures <https://docs.pytest.org/en/latest/explanation/fixtures.html>`__ are a
+great pytest tool to provide and share initial setup and prepare data for a
+test. However, they can also be a great enemy if you overuse them or if you nest
+them into multiple levels (using fixture inside a fixture).
+
+Overusing fixtures makes it quite difficult to understand what a test does and
+what data and setup does it require. This is because the information is not
+present directly in the test itself but on different place or places in the
+code. Therefore you have to jump back and forth in the code in order to
+understand what the test does. This is especially bad in testing projects like
+SSSD that has so many components.
+
+Another big downside of using fixtures is that they do not allow slight
+modifications of the setup. Most of the time, you need to write multiple tests
+for single functionality. And even though it seems logical that these tests
+share the same setup, it is most often not the case as each test usually
+requires slight modification of the overall setup. If the setup is done with
+fixtures and you need to add a new test case that requires slight modification
+you either end up duplicating the fixture code, creating more fixtures or
+refactoring the fixture and every single related test. This of course makes the
+tests harder to understand and extend and it diminishes the benefit of using
+fixtures.
+
+The SSSD test framework :doc:`lib.sssd <api>` makes the SSSD related setups quite
+easy, with just a few lines of code where everything is clear out of the box
+even without reading any documentation. Therefore there is no need to use
+fixtures.
+
+.. warning::
+
+    The general recommendation is: **Avoid using fixtures** *unless you have a
+    very good reason to use it.*
+
+Organizing test cases
+*********************
+
+Pytest allows you to write tests inside a class (starts with `Test`) or directly
+inside a module (a function starting with `test_`). Even though it might be
+logical to organize tests inside a class, it does not give you any benefit over
```

Comment:
Testclass gives the benefit of adding markers (like feature autofs eg. ) to the whole class so it can be inherited by tests not repeating the same marker for each test.

Reply:
I sent this to the team list.